datatypes/src/raster/mod.rs

```diff
@@ -15,45 +15,54 @@ pub use self::grid_bounds::{
     GridBoundingBox, GridBoundingBox1D, GridBoundingBox2D, GridBoundingBox3D,
 };
 pub use self::grid_index::{GridIdx, GridIdx1D, GridIdx2D, GridIdx3D};
+pub use self::grid_or_empty::GridOrEmpty;
 pub use self::grid_traits::{
     BoundedGrid, GridBounds, GridContains, GridIndexAccess, GridIndexAccessMut, GridIntersection,
     GridSize, GridSpaceToLinearSpace,
 };
 pub use self::grid_typed::{TypedGrid, TypedGrid2D, TypedGrid3D};
+pub use self::no_data_grid::{NoDataGrid, NoDataGrid1D, NoDataGrid2D, NoDataGrid3D};
 pub use self::operations::{blit::Blit, grid_blit::GridBlit};
-pub use self::raster_tile::{RasterTile, RasterTile2D, RasterTile3D};
+pub use self::raster_tile::{
+    BaseTile, MaterializedRasterTile, MaterializedRasterTile2D, MaterializedRasterTile3D,
+    RasterTile, RasterTile2D, RasterTile3D,
+};
 pub use self::tiling::{TileInformation, TilingSpecification, TilingStrategy};
 pub use self::typed_raster_conversion::TypedRasterConversion;
 pub use self::typed_raster_tile::{TypedRasterTile2D, TypedRasterTile3D};
+pub use self::{
+    grid_traits::ChangeGridBounds, grid_traits::GridShapeAccess, grid_traits::NoDataValue,
+};
 
 mod data_type;
 mod geo_transform;
 mod grid;
 mod grid_bounds;
 mod grid_index;
+mod grid_or_empty;
 mod grid_traits;
-
 mod grid_typed;
 mod macros_raster;
 mod macros_raster_tile;
+mod no_data_grid;
 mod operations;
 mod raster_tile;
 mod tiling;
 mod typed_raster_conversion;
 mod typed_raster_tile;
 
-pub trait Raster<D, T: Pixel, C>: SpatialBounded + TemporalBounded {
-    /// returns the grid dimension object of type D: `GridDimension`
-    fn dimension(&self) -> D;
-    /// returns the optional  no-data value used for the raster
-    fn no_data_value(&self) -> Option<T>;
-    // Checks whether the value is no data
-    fn is_no_data(&self, value: T) -> bool {
-        self.no_data_value()
-            .map_or(false, |no_data_value| value == no_data_value)
-    }
+pub trait Raster<D: GridSize, T: Pixel, C>:
```

Comment:
Maybe we should have the container as an associated type? :thinking: What do you think?

Reply:
very good idea

68824c5